src/Resolvers/Microsoft.DotNet.NativeWrapper/Interop.cs

```diff
@@ -12,36 +13,28 @@ namespace Microsoft.DotNet.NativeWrapper
     public static partial class Interop
     {
         public static readonly bool RunningOnWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
+        static readonly string HostFxrPath;
 
         static Interop()
         {
-            if (RunningOnWindows)
-            {
-                PreloadWindowsLibrary("hostfxr.dll");
-            }
+            HostFxrPath = (string)AppContext.GetData("HOSTFXR_PATH");
+            NativeLibrary.SetDllImportResolver(typeof(Interop).Assembly, HostFxrDllImportResolver);
         }
 
-        // MSBuild SDK resolvers are required to be AnyCPU, but we have a native dependency and .NETFramework does not
-        // have a built-in facility for dynamically loading user native dlls for the appropriate platform. We therefore 
-        // preload the version with the correct architecture (from a corresponding sub-folder relative to us) on static
-        // construction so that subsequent P/Invokes can find it.
-        private static void PreloadWindowsLibrary(string dllFileName)
+        static IntPtr HostFxrDllImportResolver(string libraryName, Assembly assembly, DllImportSearchPath? searchPath)
         {
-            string basePath = Path.GetDirectoryName(typeof(Interop).Assembly.Location);
-            string architecture = IntPtr.Size == 8 ? "x64" : "x86";
-            string dllPath = Path.Combine(basePath, architecture, dllFileName);
+            if (libraryName == "hostfxr")
```

Comment:
Maybe define some constant for the name, so that it is obvious that it maps to the `DllImport` declarations below?